module/os/linux/zfs/vdev_disk.c

```diff
@@ -930,37 +1138,44 @@ vdev_disk_io_start(zio_t *zio)
 		rw_exit(&vd->vd_lock);
 		zio_execute(zio);
 		return;
-	case ZIO_TYPE_WRITE:
-		rw = WRITE;
-		break;
-
-	case ZIO_TYPE_READ:
-		rw = READ;
-		break;
 
 	case ZIO_TYPE_TRIM:
 		zio->io_error = vdev_disk_io_trim(zio);
 		rw_exit(&vd->vd_lock);
 		zio_interrupt(zio);
 		return;
 
-	default:
+	case ZIO_TYPE_READ:
+	case ZIO_TYPE_WRITE:
+		zio->io_target_timestamp = zio_handle_io_delay(zio);
+		error = vdev_disk_io_rw(zio);
 		rw_exit(&vd->vd_lock);
-		zio->io_error = SET_ERROR(ENOTSUP);
-		zio_interrupt(zio);
+		if (error) {
+			zio->io_error = error;
+			zio_interrupt(zio);
+		}
 		return;
-	}
 
-	zio->io_target_timestamp = zio_handle_io_delay(zio);
-	error = __vdev_disk_physio(vd->vd_bdev, zio,
-	    zio->io_size, zio->io_offset, rw, 0);
-	rw_exit(&vd->vd_lock);
+	default:
+		/*
+		 * Getting here means our parent vdev has made a very strange
+		 * request of us, and shouldn't happen. Assert as much in
+		 * dev builds, but in production return the IO unhandled.
+		 * The pool will likely suspend anyway but that's nicer than
+		 * crashing the kernel.
+		 */
+		ASSERT(zio->io_type == ZIO_TYPE_READ ||
+		    zio->io_type == ZIO_TYPE_WRITE ||
+		    zio->io_type == ZIO_TYPE_TRIM ||
+		    zio->io_type == ZIO_TYPE_IOCTL);
```

Comment:
This looks like impossible condition. Looking on the previous cases.

Reply:
Its supposed be, to make it trip in dev, but of course `ASSERT(0)` would have just done as well. Changed to `ASSERT3S(zio->io_type, ==, -1)` to get the type into the crash output.